metrics_sdk/lib/opentelemetry/sdk/metrics/state/metric_stream.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Metrics
+      module State
+        # @api private
+        #
+        # The MetricStream class provides SDK internal functionality that is not a part of the
+        # public API.
+        class MetricStream
+          attr_reader :name, :description, :unit, :instrument_kind, :instrumentation_library, :data_points
+
+          def initialize(
+            name,
+            description,
+            unit,
+            instrument_kind,
+            meter_provider,
+            instrumentation_library
+          )
+            @name = name
+            @description = description
+            @unit = unit
+            @instrument_kind = instrument_kind
+            @meter_provider = meter_provider
+            @instrumentation_library = instrumentation_library
+
+            @data_points = {}
+            @mutex = Mutex.new
+          end
+
+          def collect(start_time, end_time)
+            @mutex.synchronize do
+              MetricData.new(
```

Comment:
I'm not clear on how instantiating this object fulfills the obligations of `collect`, but I'm definitely missing something....

Reply:
Think of it as collecting a snapshot of the metrics at a certain point in time. The metrics will accumulate between _collect calls_, so we want to capture all the information necessary to export and the current state of the data points. What's not shown here is that we will also need to configure the temporality of the metrics, so in the case of delta, after grabbing this snapshot we will drop all the datapoints. In cumulative temporality we need the datapoints and their values at the time of collection, but after collection finishes the data points on the metric stream can continue to grow, but we don't want to have that data change once it has been sent off for export. This is intended to be a simple struct to encapsulate this information distinctly from the metric stream.